### PR TITLE
refactor: Command for set delivery address

### PR DESCRIPTION
**Pull request by marlass**

### CODE REVIEW

File: feature-libs/checkout/core/facade/checkout-delivery.service.ts [hunk 1]
```diff
@@ -1,18 +1,34 @@
...
+import { combineLatest, Observable, of } from 'rxjs';
+import {
+  pluck,
+  shareReplay,
+  switchMap,
+  take,
+  tap,
+  withLatestFrom,
+} from 'rxjs/operators';
+import { CheckoutDeliveryConnector } from '..';
```
Comment (znikola) [inline feature-libs/checkout/core/facade/checkout-delivery.service.ts]: Can you please improve the import path to import directly from the file?

File: feature-libs/checkout/core/facade/checkout-delivery.service.spec.ts [hunk 2]
```diff
@@ -10,6 +10,7 @@ import {
   UserIdService,
 } from '@spartacus/core';
 import { of } from 'rxjs';
+import { CheckoutDeliveryConnector } from '..';
```
Comment (znikola) [inline feature-libs/checkout/core/facade/checkout-delivery.service.spec.ts]: Same as in the service impl, but not as important. Up to you :)

File: feature-libs/checkout/core/facade/checkout-delivery.service.ts [hunk 3]
```diff
@@ -158,11 +178,69 @@ export class CheckoutDeliveryService implements CheckoutDeliveryFacade {
     );
   }
 
+  protected createDeliveryAddressCommand = this.command?.create<Address>(
```
Comment (znikola) [inline feature-libs/checkout/core/facade/checkout-delivery.service.ts]: Going to our conversation, not sure if we want to type this property?

File: feature-libs/checkout/core/facade/checkout-delivery.service.ts [hunk 4]
```diff
@@ -158,11 +178,69 @@ export class CheckoutDeliveryService implements CheckoutDeliveryFacade {
...
+      ]).pipe(
+        take(1),
+        switchMap(([userId, cartId]) => {
+          if (
+            userId &&
+            cartId &&
+            (userId !== OCC_USER_ID_ANONYMOUS ||
+              this.activeCartService.isGuestCart()) &&
+            this.checkoutDeliveryConnector // TODO: Remove check in 5.0 when service will be required
+          ) {
```
Comment (znikola) [inline feature-libs/checkout/core/facade/checkout-delivery.service.ts]: Can we inverse the `if` check, and return `of()` immediately?

File: feature-libs/checkout/core/facade/checkout-delivery.service.ts [hunk 5]
```diff
@@ -158,11 +178,69 @@ export class CheckoutDeliveryService implements CheckoutDeliveryFacade {
...
+                  if (payload.region?.isocodeShort) {
+                    Object.assign(address.region, {
+                      isocodeShort: payload.region.isocodeShort,
+                    });
+                  }
+                  return this.setDeliveryAddress(address);
+                })
+              );
+          }
+          return of(); // TODO: should we throw error here? useful dev info?
```
Comment (znikola) [inline feature-libs/checkout/core/facade/checkout-delivery.service.ts]: Maybe we should complete the stream and return `EMPTY`?

File: feature-libs/checkout/core/facade/checkout-delivery.service.ts [hunk 6]
```diff
@@ -243,11 +322,79 @@ export class CheckoutDeliveryService implements CheckoutDeliveryFacade {
     }
   }
 
+  protected setDeliveryAddressCommand = this.command?.create<Address>(
```
Comment (znikola) [inline feature-libs/checkout/core/facade/checkout-delivery.service.ts]: The same as above - typing?

File: feature-libs/checkout/core/facade/checkout-delivery.service.ts [hunk 7]
```diff
@@ -243,11 +322,79 @@ export class CheckoutDeliveryService implements CheckoutDeliveryFacade {
...
+        take(1),
+        switchMap(([userId, cartId]) => {
+          if (
+            userId &&
+            cartId &&
+            (userId !== OCC_USER_ID_ANONYMOUS ||
+              this.activeCartService.isGuestCart()) &&
+            addressId &&
+            this.checkoutDeliveryConnector // TODO: Remove check in 5.0 when service will be required
+          ) {
```
Comment (znikola) [inline feature-libs/checkout/core/facade/checkout-delivery.service.ts]: The same as above - inverse the `if`?

File: feature-libs/checkout/core/store/actions/checkout.action.ts [hunk 8]
```diff
@@ -33,12 +33,27 @@ export const CLEAR_CHECKOUT_DELIVERY_MODE_SUCCESS =
 export const CLEAR_CHECKOUT_DELIVERY_MODE_FAIL =
   '[Checkout] Clear Checkout Delivery Mode Fail';
 
+/**
+ * @deprecated since 4.3.0. Add delivery address is handled with command in CheckoutDeliveryService.
```
Comment (znikola) [inline feature-libs/checkout/core/store/actions/checkout.action.ts:37]: Is 4.3 even planned for? 🤔

**Review by znikola — CHANGES_REQUESTED** [review]

File: feature-libs/checkout/core/store/actions/checkout.action.ts (hunk 8 above)
Comment (marlass) [reply feature-libs/checkout/core/store/actions/checkout.action.ts:37]: We will update that comments when we will know when we want to merge it :)

I assume we will have 3 minor before major

**Review by znikola — APPROVED** [review]

**Pull request merged**